shachain/store.go

```diff
@@ -0,0 +1,187 @@
+package shachain
+
+import (
+	"bytes"
+	"encoding/binary"
+	"github.com/go-errors/errors"
+	"github.com/roasbeef/btcd/wire"
+)
+
+// SecretStore is an interface which serves as an abstraction over data
+// structure responsible for effecient storing and restoring of hash secrets
+// by given indexes.
+//
+// Description: The Lightning Network wants a chain of (say 1 million)
+// unguessable 256 bit values; we generate them and send them one at a time
+// to a remote node.  We don't want the remote node to have to store all the
+// values, so it's better if they can derive them once they see them.
+type SecretStore interface {
+	// Restore function is ised to restore the previous secret by its index.
+	Restore(uint64) (*wire.ShaHash, error)
```

Comment:
I think `LookUp` is a better name here than `Restore`. Or possibly we can go with `Fetch`.